crates/ruff_linter/src/rules/flake8_pyi/rules/bad_generator_return_type.rs

```diff
@@ -0,0 +1,191 @@
+/// could change at any time. Type annotations help define a contract for a
+/// function; implementation details should not leak into that contract.
+///
+/// For example:
+/// ```python
+/// from collections.abc import AsyncGenerator, Generator
+/// from typing import Any
+///
+///
+/// class CustomIterator:
+///     def __iter__(self) -> Generator:
+///         yield from range(42)
+///
+///
+/// class CustomIterator2:
+///     def __iter__(self) -> Generator[str, Any, None]:
+///         yield from "abcdefg"
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from collections.abc import Iterator
+///
+///
+/// class CustomIterator:
+///     def __iter__(self) -> Iterator:
+///         yield from range(42)
+///
+///
+/// class CustomIterator2:
+///     def __iter__(self) -> Iterator[str]:
+///         yield from "abdefg"
+/// ```
+#[violation]
+pub struct GeneratorReturnFromIterMethod {
+    better_return_type: String,
+    method_name: String,
+}
+
+impl Violation for GeneratorReturnFromIterMethod {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let GeneratorReturnFromIterMethod {
+            better_return_type,
+            method_name,
+        } = self;
+        format!("Use `{better_return_type}` as the return value for simple `{method_name}` methods")
+    }
+}
+
+/// PYI058
+pub(crate) fn bad_generator_return_type(
+    checker: &mut Checker,
+    stmt: &ast::Stmt,
+    is_async: bool,
+    name: &str,
+    returns: Option<&ast::Expr>,
+    parameters: &ast::Parameters,
+    body: &Vec<ast::Stmt>,
+) {
+    if is_async {
+        return;
+    }
+
+    let better_return_type = match name {
+        "__iter__" => "Iterator",
+        "__aiter__" => "AsyncIterator",
+        _ => return,
+    };
+
+    let semantic = checker.semantic();
+
+    if !matches!(checker.semantic().current_scope().kind, ScopeKind::Class(_)) {
+        return;
+    }
+
+    if !parameters.kwonlyargs.is_empty()
+        || parameters.kwarg.is_some()
+        || parameters.vararg.is_some()
+    {
+        return;
+    }
+
+    if (parameters.args.len() + parameters.posonlyargs.len()) != 1 {
+        return;
+    }
+
+    let Some(returns) = returns else {
+        return;
+    };
+
+    if !semantic
+        .resolve_call_path(map_subscript(returns))
+        .is_some_and(|call_path| {
+            matches!(
+                (name, call_path.as_slice()),
+                (
+                    "__iter__",
+                    ["typing" | "typing_extensions", "Generator"]
+                        | ["collections", "abc", "Generator"]
+                ) | (
+                    "__aiter__",
+                    ["typing" | "typing_extensions", "AsyncGenerator"]
+                        | ["collections", "abc", "AsyncGenerator"]
+                )
+            )
+        })
+    {
+        return;
+    };
+
+    // `Generator`` allows three type parameters; `AsyncGenerator`` allows two.
```

Comment:
Nit: very minor (sorry!) but looks like you have extra backticks after "Generator" and "AsyncGenerator".

Reply:
you can defeat the borrow checker, but you can never defeat typos apparently